### PR TITLE
Fix invalid_scope, ensure UserPoolDomain is unique and active.

### LINKED ISSUE #415: cant open the admin app after installation
the deployment was successful but when I open the link and click on sign in this shows up
Screen Shot 2022-11-11 at 3 45 00 PM

Screen Shot 2022-11-11 at 3 42 16 PM

branch:latest main

### PR DESCRIPTION
**Pull request by PoeppingT**
This commit contains three bugfixes.
1. Tell Cognito to expect all three default OAuth scopes: openid, profile, and email.
2. Ensure that UserPoolDomain is globally unique by pulling randomness from the CFN-generated random string in the stack name. This prevents collisions across accounts when different people create environments with the same name.
3. Ensure that the CustomizeCognitoUi custom resource only acts once the UserPoolDomain has reached the ACTIVE state, since the Cognito API call to customize the hosted UI will fail otherwise.

Fixes #415 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java [hunk 1]
```diff
@@ -134,6 +137,25 @@ public Object handleRequest(Map<String, Object> event, Context context) {
         return null;
     }
 
+    protected void waitForActiveDomain(String userPoolDomain) throws Exception {
```
Comment (brtrvn) [inline resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java:140]: Prefer RuntimeException over checked exceptions

File: resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java [hunk 2]
```diff
@@ -134,6 +137,25 @@ public Object handleRequest(Map<String, Object> event, Context context) {
...
+        // Cognito cannot set UI customization until UserPoolDomain is ACTIVE
+        // ACTIVE, CREATING, DELETING, FAILED, UPDATING, UNKNOWN_TO_SDK_VERSION
+        DomainStatusType domainStatus = DomainStatusType.CREATING;
+        List<DomainStatusType> terminalStatuses = List.of(
+                DomainStatusType.ACTIVE,
+                DomainStatusType.FAILED,
+                DomainStatusType.UNKNOWN_TO_SDK_VERSION);
+        do {
+            domainStatus = cognito.describeUserPoolDomain(request -> request.domain(userPoolDomain))
```
Comment (brtrvn) [inline resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java:149]: Will we be throttled here in a way that the SDK client won't be able to deal with?

Comment (PoeppingT) [reply resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java:149]: Good point, I'll implement some throttling backoff.

File: resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java (hunk 1 above)
Comment (PoeppingT) [reply resources/custom-resources/customize-cognito-ui/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CustomizeCognitoUi.java:140]: Can do.

**Review by brtrvn — APPROVED** [review]
Note that updating an existing SaaS Boost environment using the `install.sh` script will fail with this PR because of CloudFormation limitations with Cognito user pool domains. To apply this fix to an existing environment:
1. Either run update from the installer and wait for the CloudFormation stack update to fail, or run `update.sh` in the `resources/custom-resources/customize-cognito-ui`
2. Comment out the `UserPoolDomain` resource and the `CustomizeCognitoUi` resource in the `saas-boost-idp.yaml` template and copy the modified template to the SaaS Boost artifacts bucket for your environment
3. Run an update stack on the base SaaS Boost stack in CloudFormation
4. Remove the comments from the `saas-boost-idp.yaml` template and copy the modified template to the SaaS Boost artifacts bucket for your environment
5. Run an update stack a 2nd time on the base SaaS Boost stack in CloudFormation

**Pull request merged**